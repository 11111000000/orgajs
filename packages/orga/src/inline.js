import Node from './node'

const LINK_PATTERN = /(.*?)\[\[([^\]]*)\](?:\[([^\]]*)\])?\](.*)/m; // \1 => link, \2 => text

const PRE = `[\\s\\({'"^]`
const POST = `[\\s-\\.,:!?'\\)}$]`
const BORDER = `[^,'"\\s]`

function markup(marker) {
  return RegExp(`(.*?${PRE})${marker}(${BORDER}.+?${BORDER})${marker}(${POST}.*)`, 'm')
}

function parse(text) {
  text = _parse(LINK_PATTERN, text, (captures) => {
    return new Node(`link`).with({ path: captures[0], desc: captures[1] })
  })


  const markups = [
    { name: `bold`, pattern: markup(`\\*`) },
    { name: `verbatim`, pattern: markup(`=`) },
    // { name: `italic`, pattern: `/` }, // TODO: deal with links first
    { name: `strike-through`, pattern: markup(`\\+`) },
    { name: `underline`, pattern: markup(`_`) },
    { name: `code`, pattern: markup(`~`) },
  ]
  for (const { name, pattern } of markups) {
    text = _parse(pattern, text, (captures) => { return new Node(name).with({ value: captures[0] }) })
  }
  return text
}


function _parse(pattern, text, post) {
  if (typeof text === `string`) {
    var m = pattern.exec(text)
    if (!m) return [new Node(`text`).with({ value: text })]
    m.shift()
    let before = m.shift()
    let after = m.pop()
    var nodes = []
    if ( before.length > 0 ) {
      nodes.push(new Node(`text`).with({ value: before }))
    }
    if (m.length > 0) {
      nodes.push(post(m))
      // nodes.push(new Node(name).with({ value: match }))
    }
    if (after) {
      nodes = nodes.concat(_parse(pattern, after, post))
    }
    return nodes
  }

  if (Array.isArray(text)) {
    return text.reduce((all, node) => {
      if (node.hasOwnProperty(`type`) && node.type != `text`) {
        return all.concat(node)
      }
      return all.concat(_parse(pattern, node, post))
    }, [])
  }

  if (typeof text.value === `string`) {
    return _parse(pattern, text.value, post)
  }
  return undefined
}

module.exports = {
  parse,
}
