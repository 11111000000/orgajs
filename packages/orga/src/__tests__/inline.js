import { parse } from '../inline'
import Node from '../node'


describe('Inline Parsing', () => {
  it('recon single emphasis', () => {

    const str = 'hello *world*, welcome to *org-mode*. Have a nice *day*.'
    expect(parse(str)).toMatchSnapshot()
  })

  it('recon mixed emphasis', () => {

    const str = `
[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~, the round pegs in the +round+ square holes...
`
    expect(parse(str)).toMatchSnapshot()
  })

  it('recon emphasises at different locations', () => {
    expect(parse(`one *two* three`)).toMatchSnapshot()
    expect(parse(`*one* two three`)).toMatchSnapshot()
    expect(parse(`one two *three*`)).toMatchSnapshot()
  })
})
