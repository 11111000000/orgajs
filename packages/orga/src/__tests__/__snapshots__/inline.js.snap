// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline Parsing recon emphasises at different locations 1`] = `
Array [
  Object {
    "children": Array [],
    "type": "text",
    "value": "one ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "two",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": " three",
  },
]
`;

exports[`Inline Parsing recon emphasises at different locations 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "one",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": " two three",
  },
]
`;

exports[`Inline Parsing recon emphasises at different locations 3`] = `
Array [
  Object {
    "children": Array [],
    "type": "text",
    "value": "one two ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "three",
      },
    ],
    "type": "bold",
  },
]
`;

exports[`Inline Parsing recon link 1`] = `
Array [
  Object {
    "children": Array [],
    "type": "text",
    "value": "hello ",
  },
  Object {
    "children": Array [],
    "desc": undefined,
    "type": "link",
    "uri": Object {
      "location": "image/logo.png",
      "protocol": "file",
      "raw": "image/logo.png",
    },
  },
]
`;

exports[`Inline Parsing recon link 2`] = `
Array [
  Object {
    "children": Array [],
    "type": "text",
    "value": "hello ",
  },
  Object {
    "children": Array [],
    "desc": "logo",
    "type": "link",
    "uri": Object {
      "location": "image/logo.png",
      "protocol": "file",
      "raw": "image/logo.png",
    },
  },
]
`;

exports[`Inline Parsing recon mixed emphasis 1`] = `
Array [
  Object {
    "children": Array [],
    "desc": "Here's",
    "type": "link",
    "uri": Object {
      "location": "//github.com/xiaoxinghu/orgajs",
      "protocol": "https",
      "raw": "https://github.com/xiaoxinghu/orgajs",
    },
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": " to the ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "crazy",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": " ones, the ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "misfits",
      },
    ],
    "type": "italic",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ", the ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "rebels",
      },
    ],
    "type": "underline",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ", the ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "troublemakers",
      },
    ],
    "type": "code",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ", the round pegs in the ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "round",
      },
    ],
    "type": "strikeThrough",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": " square holes...",
  },
]
`;

exports[`Inline Parsing recon single emphasis 1`] = `
Array [
  Object {
    "children": Array [],
    "type": "text",
    "value": "hello ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "world",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ", welcome to ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "org-mode",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ". Have a nice ",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "parent": [Circular],
        "type": "text",
        "value": "day",
      },
    ],
    "type": "bold",
  },
  Object {
    "children": Array [],
    "type": "text",
    "value": ".",
  },
]
`;
